### ISSUE TITLE
Current cspSsrNonce plugin can result in security problem in some cases

### ISSUE THREAD

**Issue by maple3142** [Baroshem/nuxt-security#594]
### Environment

```shell
- Operating System: Linux
- Node Version:     v22.11.0
- Nuxt Version:     3.14.1592
- CLI Version:      3.17.2
- Nitro Version:    2.10.4
- Package Manager:  yarn@1.22.22
- Builder:          -
- User Config:      default
- Runtime Modules:  nuxt-security@2.1.5
- Build Modules:    -
```

### Nuxt Security Version

v2.1.5

### Default setup used?

Yes, the bug happens even if the security option is not customized

### Security options

```ts

```

### Reproduction

https://github.com/maple3142/nuxt-security-nonce-bug-repro

1. `yarn install && yarn build`
2. `node .output/server/index.mjs`
3. Visit `http://localhost:3000/?statusMessage=1&html=%3Cdiv+id=%22%3Cscript%3E%3Cscript%3Ealert(origin)%3C/script%3E%22%3E%3C/div%3E` from browser

### Description

The browser would pop an alert when visiting that url (so XSS happens here), despite https://github.com/maple3142/nuxt-security-nonce-bug-repro/blob/master/pages/index.vue does sanitize the incoming html using [DOMPurify](https://github.com/cure53/DOMPurify).

The main problem is that nuxt security incorrectly add nonces to script tags using simple string replacement here: https://github.com/nuxt-modules/security/blob/70cd67a4690b53e9167c210e3e020faed901355d/src/runtime/nitro/plugins/40-cspSsrNonce.ts#L53-L73 . And `v-html` can result in unescaped `<script>` to appear in rendered html in SSR mode.

I think nuxt-security should not process html in from v-html, and should not naively use string replacement to do that because it may accidentally replace fake script tags inside element attributes.

### Logs

```shell

```

**Comment by vejja** [comment]
Hey @maple3142 
I'm still turning my head around the potential implications of this, but your example is not a bug.

What you are doing here is not XSS. You are SSR-ing a dynamic page.
Your setup uses an SSR server, and you are instructing this server to generate a page that contains a `<script>` element.
Your server is following your instructions faithfully, there is nothing wrong about this.

XSS happens when a page is modified on the _client-side_.
The CSP security model prevents client-side injections, but assumes that anything delivered by the server is valid.
It is therefore the server's responsibility to generate safe content. 

Using `v-html` with a random user-provided input string is a security hole that you are opening yourself on the server side.
CSP will not prevent the script from executing, because it assumes that your server is safe.

That being said, the specifics of your example rely on using `DOMPurify` to sanitize the user input. I'm not knowledgeable about this library so I won't be able to comment specifically. However it looks like the user input is not sanitized from potential `script` tags.

**Comment by maple3142** [comment]
The problem here is that DOMPurify does it job correct, because the `<script>` is inside an element's attribute, which does not need to be escaped.

For example, the following html is completely safe:

```html
<div id="<script>">content</div>
```

so DOMPurify will left it as is. If that script tag is not inside that attribute then DOMPurify will remove it as you would expect.

The problem here is that nuxt-security does a simple string replacement to add nonce to script tags, without considering whether it is actually a script tag or not. In this case, nuxt-security does it on `<script>` on text inside a html attribute.

You can try to uninstall nuxt-security from my example repo and try again the url that triggers XSS, and you will see it does not happen anymore.

**Comment by vejja** [comment]
I get this part:
> The problem here is that nuxt-security does a simple string replacement to add nonce to script tags, without considering whether it is actually a script tag or not.

Which is a valid comment.

However I still don't understand where is the XSS ?

**Comment by maple3142** [comment]
The input to DOMPurify is the following HTML:

```html
<div id="<script><script>alert(origin)</script>"></div>
```

and after DOMPurify's sanitization, it becomes:

```html
<div id="<script><script>alert(origin)</script>"></div>
```

(no changes)

When this html is rendered by SSR, it **does not** result in XSS because it is not a script tag.

But if nuxt-security is installed, it becomes:

```html
<div id="<script nonce="NONCE"><script nonce="NONCE">alert(origin)</script>"></div>
```

and this does result in the `alert` being executed by browsers, which is an XSS introduced by nuxt-security.

A fix for this is to modify cspSsrNonce to add nonce by properly parsing html so it no longer add nonce for `<script>` text inside attributes.

**Comment by vejja** [comment]
Ok, I get it now

@GalacticHypernova would you have an idea for this ?
@maple3142 actually has a valid point in his description above

The issue is that:
```html
<div id="<script><script>alert(origin)</script>">
</div>
```
Gets transformed into:
```html
<!-- here is what happens below -->
<!-- the opening quote of the nonce actually closes the id attributed -->
<!-- then the NONCE value is interpreted as an attribute -->
<!-- finally the closing quote is ignored -->
<div id="<script nonce=" NONCE "> 
  <!-- this causes a second script block to be inserted !-->
  <!-- which also has a nonce, so it gets executed -->
  <script nonce="NONCE"> 
    alert(origin)
  </script>
  ">
</div>
```

Formatting, block indenting, extra spaces and comments are from me to increase legibility of what happens.
The attack is based on quote mismatch inside the `id` attribute. 
It makes the assumption that the server :
1. provides a public route that accepts the injection of any html into the DOM
2. can do so because it relies on a 3rd-party library to sanitize user input (here, DOMPurify)

**Comment by GalacticHypernova** [comment]
The issue I see in this is that technically, this goes under the assumption that we purposefully try to make a non-standard ID. Given that this is in the developer's control, it's worth noting the improbability that someone would actually name it this way. After all, an ID is primarily used to link to other content on the same page (#id), and I don't think it would be very often that such id would be introduced. We can, of course, modify the regex to capture that, but it's not an XSS from nuxt-security, it's just bad practice from the developer.

**Comment by GalacticHypernova** [comment]
If it turns out to be too resource-intesive to the point where it causes noticeable slowdown, I believe we should really think whether or not this is necessary. 

If someone would really want to make an XSS in their website, they will be able to even with the strictest settings and behavior imaginable. 

Besides the security of the tools in use, there's also the responsibility of the developer to follow the web specs and best practices, including security.

**Comment by maple3142** [comment]
Just a note that this has nothing to do with the specific attribute `id`, as it can happen with any html attributes like `href`, `alt` or whatever.

Also, in normal usecase like `<img :src="user_url" :alt="user_alt">` this attack would not apply, because vue will encode `<>` characters in attributes.

The main problem about this issue is when website want to allow user supplied html to some extent (or markdown/bbcode/whatever), so they apply html sanitizer like DOMPurify to do that and put it into `v-html`. I don't think there are other cases where this can apply in the context of nuxt.

**Comment by vejja** [comment]
> Just a note that this has nothing to do with the specific attribute `id`, as it can happen with any html attributes like `href`, `alt` or whatever.

Yes, understood

> Also, in normal usecase like `<img :src="user_url" :alt="user_alt">` this attack would not apply, because vue will encode `<>` characters in attributes.

Exact, although we don't run at the Vue level - we only interact after the final HTML generation step so I don't think we need to worry about this

> 
> The main problem about this issue is when website want to allow user supplied html to some extent (or markdown/bbcode/whatever), so they apply html sanitizer like DOMPurify to do that and put it into `v-html`. I don't think there are other cases where this can apply in the context of nuxt.

Yes, but that's the main issue technically for us. At the time we interact, we don't have the information anymore whether this was generated from a `v-html` or something else

**Comment by vejja** [comment]
> If it turns out to be too resource-intesive to the point where it causes noticeable slowdown, I believe we should really think whether or not this is necessary.
> 
> If someone would really want to make an XSS in their website, they will be able to, even with the strictest settings and behavior imaginable.
> 
> Besides the security of the tools in use, there's also the responsibility of the developer to follow the web specs and best practices, including security.

Agree again.

But the issue I'm seeing here is that _we_ are the ones injecting the quotes that break the original HTML

**Comment by GalacticHypernova** [comment]
> But the issue I'm seeing here is that _we_ are the ones injecting the quotes that break the original HTML

True, but then again, that's just in a scenario that will pretty much never happen.

> Just a note that this has nothing to do with the specific attribute id, as it can happen with any html attributes like href, alt or whatever.

I mean, what legitimate use case requires an attribute, any attribute, to be a script tag?

If the developer allows people to add custom script tags to a page, that's an XSS waiting to happen. Same with styles, like key logging.  And not to mention the real life scenario of moderating it if someone can just add an inappropriate background image or a script that navigates to it. And that, as I stated previously, would be bad practice regardless of the tools used.

This isn't to say I won't review it and attempt to fix it of course, I'm just saying that if the fix happens to be unproportionally expensive, then we should likely pick and choose our battles and go with the solution that ultimately has real world impact.

**Comment by maple3142** [comment]
> I mean, what legitimate use case requires an attribute, any attribute, to be a script tag?

Not much I think, the only possible scenario I can think of is an image of HTML code, so its `alt` would contain HTML content to cater to screenreader users. However, this kind of content is still safe by itself so sanitizers like DOMPurify does keep it as it so such isseus definitely happen in real world.

> If the developer allows people to add custom script tags to a page

I think I have to reclarify that in the scenario here, the developer **does not** allows people to add custom script tags to a page by using DOMPurify. It is just text that happens to be HTML content are embedded in a HTML attribute, which is safe so there is no reason to ban that.

I think it is fair to expect that **if a nuxt app does not have security vulnerability, then adding nuxt-security to it should not introduce one**. (at least in default configuration)

> if the fix happens to be unproportionally expensive

Yeah, this is also what I would expect because the "proper" fix I have imagined would be using a proper HTML parser, but this might be expensive in terms of performance. Another possible fix would be disabling nonce based CSP by default and add a warning about this in docs, because I believe `self` and `strict-dynamic` CSP should be sufficient to most normal nuxt apps. Of course there might be some cases that I don't know isn't covered by these two CSP directives.

**Comment by GalacticHypernova** [comment]
> only possible scenario I can think of is an image of HTML code, so its `alt` would contain HTML content to cater to screenreader users

Considering you can also phrase it as "html script tag responsible for xyz", it would still be bad practice to make it in the format of an actual script tag, wouldn't it? I'm also unsure whether screen readers even actually pick up <> as readable characters.


> I think I have to reclarify that in the scenario here, the developer **does not** allows people to add custom script tags to a page by using DOMPurify.

Yes, I didn't mean in this context you referred to, my apologies if it seemed that way. As I stated in a previous message, if it's a script tag in an attribute, that's bad practice because there is definitely a better way of doing it. I was saying that if it **isn't** the case, then the only other affected use case would be developers that allow people to add custom scripts, and that's an XSS waiting to happen regardless.

> Another possible fix would be disabling nonce based CSP by default and add a warning about this in docs, because I believe `self` and `strict-dynamic` CSP should be sufficient to most normal nuxt apps. Of course there might be some cases that I don't know isn't covered by these two CSP directives.

That's a slight problem, because `strict-dynamic` relies on nonces and hashes. It allows you to specify a nonce for the "root" script, that all other scripts loaded from it would automatically be accepted.

**Comment by vejja** [comment]
> Yeah, this is also what I would expect because the "proper" fix I have imagined would be using a proper HTML parser, but this might be expensive in terms of performance. 

Absolutely. As a matter of fact, in previous versions we used to rely on Cheerio to parse HTML.
But the performance issues were terrible. The latency impact was huge and we were hitting memory limits in worker environments.
That’s why we moved to regex-based parsing.

**Comment by GalacticHypernova** [comment]
Hey!

I sincerely apologize for the amount of time it took me to come up with a solution to this.

Right about the time I made the previous iteration, I've gotten extremely busy with my personal life and had practically no time to attend to this properly.

I've made a PR to address this issue #658
